packages/core/src/main/notification/NotificationService.ts

```diff
@@ -145,26 +124,30 @@ export class NotificationService extends EventEmitter {
 
   public async handleNotificationStream(notificationHandler: NotificationHandler): Promise<void> {
     const notifications = await this.getAllNotifications();
-    for (const notification of notifications) {
-      await notificationHandler(notification, PayloadBundleSource.NOTIFICATION_STREAM).catch(error =>
-        this.logger.error(error),
-      );
+    for (const [index, notification] of notifications.entries()) {
+      await notificationHandler(notification, PayloadBundleSource.NOTIFICATION_STREAM, {
+        done: index + 1,
+        total: notifications.length,
+      }).catch(error => this.logger.error(error));
     }
   }
 
-  public readonly handleNotification: NotificationHandler = async (
+  public async *handleNotification(
     notification: Notification,
     source: PayloadBundleSource,
-  ): Promise<void> => {
+  ): AsyncGenerator<HandledEventPayload> {
     for (const event of notification.payload) {
-      let data: PayloadBundle | void;
-
+      this.logger.log(`Handling event of type "${event.type}" for notification with ID "${notification.id}"`, event);
       try {
-        this.logger.log(`Handling event of type "${event.type}" for notification with ID "${notification.id}"`, event);
-        data = await this.handleEvent(event, source);
+        const data = await this.handleEvent(event, source);
         if (!notification.transient) {
+          // keep track of the last handled notification for next time we fetch the notification stream
           await this.setLastNotificationId(notification);
         }
+        yield {
+          ...data,
+          mappedEvent: data.mappedEvent ? this.cleanupPayloadBundle(data.mappedEvent) : undefined,
+        };
```

Comment:
Instead of sending events, the method is now a generator function that warns its parent when an event has been processed. 
This way we have a clear parent>child relationship and the flow is easier to follow